### PR TITLE
Adding Occlusion

### PR DESCRIPTION
**Pull request by vivekmig**
Differential Revision: D18670640

This PR adds Occlusion, a variant of feature ablation which ablates a contiguous window which slides along the inputs in each dimension based on a given stride. For inputs contained in multiple windows, the attribution is taken as the average of the ablation differences for each of the containing windows.

### CODE REVIEW

File: captum/attr/_core/occlusion.py [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+        # Construct counts, defining number of steps to make of occlusion block in
+        # each dimension.
+        shift_counts = []
+        for i, inp in enumerate(formatted_inputs):
+            current_shape = np.subtract(inp.shape[1:], occlusion_shapes[i])
+            shift_counts.append(
+                tuple(
+                    np.add(np.ceil(np.divide(current_shape, strides[i])).astype(int), 1)
+                )
+            )
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/occlusion.py]: Good point, I think we should probably assert and raise an exception in this case. Even if the shift count was 1, it wouldn't cover the entire second dimension, so we would have missing attributions.

File: captum/attr/_core/occlusion.py [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+        assert isinstance(strides, tuple) and len(strides) == len(
+            formatted_inputs
+        ), "Strides must be provided for each input tensor."
+
+        # Construct occlusion blocks
+        if not isinstance(occlusion_shapes[0], tuple):
+            occlusion_shapes = (occlusion_shapes,)
+        assert len(occlusion_shapes) == len(
+            formatted_inputs
+        ), "Must provide occlusion dimensions for each tensor."
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/occlusion.py]: Although it is formatting, which is mostly in common, this is pretty specific to occlusion, so not sure if it'll be confusing to have it in common? The same might be true of strides, but I guess there could be more use cases of strides? I can also add it as a helper method in occlusion, what do you think?

[the remark this answers is not in the document]
Comment (NarineK) [reply captum/attr/_core/occlusion.py]: I think we can still move it there and put a comment saying that it is used by occlusion this could be used also by other image sliding window algorithms ? 
Maybe it would be even better to call, `silding_window_shapes` ? Will it be more understandable ?

Ultimately, we could move formatting from  `common.py` into `common_formatting.py` to keep it more clear and avoid common.py becoming too large.

File: captum/attr/_core/occlusion.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (NarineK) [reply captum/attr/_core/occlusion.py]: Yeah, since it is always doing plus one we might need to check that too. For example in this case:
```
formatted_inputs = (torch.rand(1, 3, 255, 255), )
occlusion_shapes = ((2, 4, 4), )
strides = (3,)
```
I'm still getting `[(2, 85, 85)]`

[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/occlusion.py]: Yes, so in this case, I intended for it to return 2 for the first dimension to ensure that it covers the full range. This results in the final patch / occlusion block to be cut off, but includes the entire range. I have a note about this in the documentation.

[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/occlusion.py]: Sorry, first misread the values. This is the same case as above right? With a stride of 3, index 2 of dimension 1 is still not covered?

File: captum/attr/_core/occlusion.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (NarineK) [reply captum/attr/_core/occlusion.py]: In general I think that it is good to describe that this is similar to sliding window because occlusion term is not very well known. When we search for it, google redirects to dentist's web sites :D

File: captum/attr/_core/occlusion.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (NarineK) [reply captum/attr/_core/occlusion.py]: yes, it is similar case: the only difference is that the stride length is equal to the 1st dimension(count starting from 0) of input. since stride length is 3, I expect to get: `[(1, 85, 85)]` instead of `[(2, 85, 85)]`, is that right ?

[the remark this answers is not in the document]
Comment (NarineK) [reply captum/attr/_core/occlusion.py]: Also, shouldn't `shift_counts` be `shift_counts=(1,33,33)` ?

[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/occlusion.py]: Hm, I think the 85, 85, is correct? The 85 makes sense since stride is 3, so last shift starts at 84*3 = 252. The extra 1 counts the origin transformation. The first dimension is off since the stride is >= (input dimension - occlusion dimension), which causes skipped regions. Will detect and raise an exception in that case, so hopefully should resolve this issue!

File: captum/attr/_core/occlusion.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+            isinstance(strides, tuple) and len(strides) == 1
+        ):
+            strides = (strides,)
+        assert isinstance(strides, tuple) and len(strides) == len(
+            formatted_inputs
+        ), "Strides must be provided for each input tensor."
+
+        # Construct occlusion blocks
+        if not isinstance(occlusion_shapes[0], tuple):
+            occlusion_shapes = (occlusion_shapes,)
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/occlusion.py]: Good point, will add a check for this!

File: captum/attr/_core/occlusion.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (NarineK) [reply captum/attr/_core/occlusion.py]: Yes, 85, 85 are correct. My point was for 2 in (2, 85, 85)

[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/occlusion.py]: Oh ok, makes sense, sounds good!

**Comment by NarineK** [comment]
Forgot to mention, can we also add a short description in this PR. OSS users cannot access task description.

File: captum/attr/_core/occlusion.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+    def _occlusion_mask(
+        self, input, feature_number, occlusion_tensor, strides, shift_counts
+    ):
+        current_total = feature_number
+        current_index = []
+        for i, val in enumerate(shift_counts):
+            stride = strides[i] if isinstance(strides, tuple) else strides
+            current_index.append((current_total % val) * stride)
+            current_total = current_total // val
+
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/occlusion.py]: I think it should just be constructing for the particular input tensor, ablation generator indexes shift_counts (and any kwarg that is a tuple), so what gets passed here should just be the shift counts for that particular tensor. I'll try to rename the variables to make this clearer.

**Comment by vivekmig** [comment]
Thanks for the review @NarineK! Addressed all comments and added checks and tests for extreme sliding window shapes and strides, let me know if you notice any other issues.

File: tests/attr/test_occlusion.py [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+        occ = Occlusion(net)
+        # Check error when too few sliding window dimensions
+        with self.assertRaises(AssertionError):
+            _ = occ.attribute(inp, sliding_window_shapes=((1, 2),), target=0,)
+
+        # Check error when too many sliding window dimensions
+        with self.assertRaises(AssertionError):
+            _ = occ.attribute(
+                (inp, inp), sliding_window_shapes=((1, 1, 2), (1, 1, 1, 2)), target=0,
+            )
```
Comment (NarineK) [inline tests/attr/test_occlusion.py]: Thank you for adding this test cases

File: captum/attr/_core/occlusion.py [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+        """
+        remaining_total = ablated_feature_num
+        current_index = []
+        for i, shift_count in enumerate(shift_counts):
+            stride = strides[i] if isinstance(strides, tuple) else strides
+            current_index.append((remaining_total % shift_count) * stride)
+            remaining_total = remaining_total // shift_count
+
+        remaining_padding = np.subtract(
+            expanded_input.shape[2:], np.add(current_index, sliding_window_tsr.shape)
```
Comment (NarineK) [inline captum/attr/_core/occlusion.py]: Can this lead to negative paddings ?

File: captum/attr/_core/occlusion.py [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+        We then need to compute the padding required after the window in each
+        dimension, which is equal to the total input dimension minus the sliding
+        window dimension minus the (left) shift amount. We construct the
+        array pad_values which contains the left and right pad values for each
+        dimension, in reverse order of dimensions, starting from the last one.
+
+        Once these padding values are computed, we pad the sliding window tensor
+        of 1s with 0s appropriately, which is the corresponding mask,
+        and the result will match the input shape.
+        """
```
Comment (NarineK) [inline captum/attr/_core/occlusion.py]: Thank you for adding the description!

File: captum/attr/_core/occlusion.py [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+                            are provided to forward_func in order following the
+                            arguments in inputs.
+                            For a tensor, the first dimension of the tensor must
+                            correspond to the number of examples. For all other types,
+                            the given argument is used for all forward evaluations.
+                            Note that attributions are not computed with respect
+                            to these arguments.
+                            Default: None
+                ablations_per_eval (int, optional): Allows multiple occlusions
+                            to be processed simultaneously in one call to forward_fn.
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/occlusion.py]: This is a little confusing since by default, each occlusion is a separate forward evaluation, and  only when this is > 1, multiple occlusions are included in the same evaluation 
(essentially 'simultaneous'). In the DataParallel case, the multiple occlusions in one batch are divided among the devices. Will reword this to make it clearer, thanks!

[the remark this answers is not in the document]
Comment (NarineK) [reply captum/attr/_core/occlusion.py]: I see, makes sense. In that case maybe we could say `multiple occlusions included in one batch` ?

[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/occlusion.py]: Sounds good, will do!

File: captum/attr/_core/occlusion.py [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+            >>> # SimpleClassifier takes a single input tensor of size Nx4x4,
+            >>> # and returns an Nx3 tensor of class probabilities.
+            >>> net = SimpleClassifier()
+            >>> # Generating random input with size 2 x 4 x 4
+            >>> input = torch.randn(2, 4, 4)
+            >>> # Defining FeatureAblation interpreter
+            >>> ablator = Occlusion(net)
+            >>> # Computes occlusion attribution, ablating each 1x3x3 patch,
+            >>> # shifting in each direction by the default of 1.
+            >>> attr = ablator.attribute(input, target=1, sliding_window_shapes=(1,3,3))
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_core/occlusion.py]: Good catch, thanks!

File: captum/attr/_core/occlusion.py (hunk 6 above)
Comment (vivekmig) [reply captum/attr/_core/occlusion.py]: Yes, this is intentional, since passing negative padding splices the corresponding dimension, which is used when the sliding window tensor needs to be cropped at edges.
